add --output-logs= option to turbo CLI

chelkyl opened the pull request:
* Adds `--output-logs` option with a few mode options to control output logs of tasks run by `turbo run build`
* `full` is default and shows all tasks' process output
* `hash-only` only outputs the hash line which has `cache <miss|hit>, <executing|replaying output> <hash>`
* `new-only` shows all process output for tasks that have not been cached and only the hash line for cached tasks
* `none` hides all tasks' process output

Usage:
`turbo run build --output-logs=new-only`

Without --output-logs or --output-logs=full
![image](https://user-images.githubusercontent.com/14041823/158048737-86cd641e-c704-4e03-a78b-0b0d7dc76361.png)

With --output-logs=hash-only
![image](https://user-images.githubusercontent.com/14041823/158048632-701a5217-a6c0-4e3a-b671-5f111288bd40.png)

With --output-logs=new-only
![image](https://user-images.githubusercontent.com/14041823/158048683-8568996a-4fac-4328-b485-34ce1b6edc84.png)

With --output-logs=none
![image](https://user-images.githubusercontent.com/14041823/158048759-174afbaa-261f-428f-a798-e340ab449808.png)

Closes #364

dcherman commented:
As a thought, can we also add `turbo logs <hash>` as a way to re-play logs from cached builds?  That way if you do choose to use this flag, you still have a way of inspecting the logs from the completed builds if needed since you'd have the hashes.

weyert commented:
I am wondering if something like similar to `docker's  `--progress` flag makes more sense here?
> --progress | auto | Set type of progress output (auto, plain, tty). Use plain to show container output

jlarmstrongiv commented:
Wow, this is exactly what I need 😄 can’t wait to see it merged

I think @dcherman’s idea is an interesting.  All logs are saved in the `node_modules/.cache/turbo/<hash>/.turbo/turbo-<command>.log` if you need to find it in the meantime.

chelkyl commented:
> I am wondering if something like similar to `docker's `--progress` flag makes more sense here?
> 
> > --progress | auto | Set type of progress output (auto, plain, tty). Use plain to show container output

@weyert That does make a lot of sense! Refactoring.

chelkyl commented:
> As a thought, can we also add `turbo logs <hash>` as a way to re-play logs from cached builds? That way if you do choose to use this flag, you still have a way of inspecting the logs from the completed builds if needed since you'd have the hashes.

@dcherman Great idea! I'll see what can be done and what other features this command could have.

weyert commented:
Nice, liking this. How can we progress this PR? @gsoltis

gsoltis commented:
I like the idea, but I'm a little concerned about the flag name and option values. As I understand it, this is intended to control the behavior of logging when a cache hit occurs. The various options are "replay the logs", "don't replay the logs", and "replay just the first line that shows a cache hit". Does that seem like an accurate summary?

`--progress` on its own doesn't, at least to me, suggest logs behavior. Maybe there's another term we could use? Something like `--cached-logs=[full|silent|name-only]` or something like that? Definitely open to suggestions on this.

chelkyl commented:
> I like the idea, but I'm a little concerned about the flag name and option values. As I understand it, this is intended to control the behavior of logging when a cache hit occurs. The various options are "replay the logs", "don't replay the logs", and "replay just the first line that shows a cache hit". Does that seem like an accurate summary?

@gsoltis Previously, the `--progress` option only had 2 modes, "replay the logs" and  "replay just the first line that shows a cache hit".

> `--progress` on its own doesn't, at least to me, suggest logs behavior. Maybe there's another term we could use? Something like `--cached-logs=[full|silent|name-only]` or something like that? Definitely open to suggestions on this.

It was based on what docker provides and I wanted to keep it to a single word but I agree that it could be clearer! I have renamed it to `--output-logs` and added more functionality to control both cached and non-cached task output.
See the updated description and let me know what you think!

gsoltis reviewed (changes requested):
This is coming along!

Can you also resolve any merge conflicts?

chelkyl commented on cli/internal/run/run.go:
```diff
@@ -427,9 +432,15 @@ type RunOptions struct {
-	dryRun     bool
-	dryRunJson bool
+	only bool
+	// Task logs output modes (cached and not cached tasks):
```
Done! Let me know if you prefer other names/values

chelkyl commented on cli/internal/run/run.go:
```diff
@@ -427,9 +432,15 @@ type RunOptions struct {
+	// full - show all,
+	// hash - only show task hash,
+	// none - show nothing
+	cachedOutputLogsMode    string
```
Agreed! Done

chelkyl commented on cli/internal/run/run.go:
```diff
@@ -427,9 +432,15 @@ type RunOptions struct {
+	// hash - only show task hash,
+	// none - show nothing
+	cachedOutputLogsMode    string
+	notCachedOutputLogsMode string
```
Agreed! Done

chelkyl commented on cli/internal/run/run.go:
```diff
@@ -559,6 +572,24 @@ func parseRunArgs(args []string, output cli.Ui) (*RunOptions, error) {
 				includDepsSet = true
 			case strings.HasPrefix(arg, "--only"):
 				runOptions.only = true
+			case strings.HasPrefix(arg, "--output-logs"):
```
Great catch! Fixed

chelkyl commented on cli/internal/run/run.go:
```diff
@@ -559,6 +572,24 @@ func parseRunArgs(args []string, output cli.Ui) (*RunOptions, error) {
 				runOptions.only = true
+			case strings.HasPrefix(arg, "--output-logs"):
+				outputLogsMode := arg[len("--output-logs="):]
+				if len(outputLogsMode) > 0 {
```
That makes sense to me! Done

chelkyl commented:
Thanks!
